packages/fxa-auth-server/scripts/cleanup-partial-firestore-customers.ts

```diff
@@ -0,0 +1,231 @@
+  private cloudTaskQueue: PQueue;
+
+  constructor(
+    private batchSize: number,
+    private dryRun: boolean,
+    rateLimit: number
+  ) {
+    this.firestore = Container.get<Firestore>(AuthFirestore);
+    this.log = Container.get(AuthLogger);
+    this.config = Container.get(AppConfig);
+    this.accountDeleteManager = Container.get(AccountDeleteManager);
+    this.accountManager = Container.get(AccountManager);
+    this.cloudTaskQueue = new PQueue({ concurrency: rateLimit });
+  }
+
+  /**
+   * Fetches subscriptions from Firestore paginated by batchSize
+   * @param startAfter ID of the last element of the previous batch for pagination
+   * @returns A list of subscriptions from firestore
+   */
+  private async fetchFirestoreCustomersUid(
+    startAfter: QueryDocumentSnapshot | undefined
+  ) {
+    this.log.debug('CleanupFirestoreHelper.fetchFirestoreCustomersUid.start', {
+      startAfterDocId: startAfter?.id,
+    });
+
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const customerCollection = `${collectionPrefix}customers`;
+    const customerSnap = startAfter
+      ? await this.firestore
+          .collectionGroup(customerCollection)
+          .orderBy(FieldPath.documentId())
+          .startAfter(startAfter)
+          .limit(this.batchSize)
+          .get()
+      : await this.firestore
+          .collectionGroup(customerCollection)
+          .orderBy(FieldPath.documentId())
+          .limit(this.batchSize)
+          .get();
+
+    // Customer document ids are the FxA Account UID
+    const uids = customerSnap.docs.map((doc) => doc.id);
+    const newStartAfter = customerSnap.docs.at(-1);
+
+    return { uids, startAfter: newStartAfter };
+  }
+
+  private async getUidsForDelete(uids: string[]) {
+    this.log.debug('CleanupFirestoreHelper.getUidsForDelete.start', {
+      nrOfUids: uids.length,
+    });
+
+    // Fetch all accounts for uids
+    const accounts = await this.accountManager.getAccounts(uids);
+    const accountUids = accounts.map((account) =>
+      uuidTransformer.from(account.uid)
+    );
+    // Filter uids where account does not exist
+    // Uids where account doesn't exist, should be deleted
+    const uidsForDelete = uids.filter(
+      (uid) => !accountUids.find((accountId) => accountId === uid)
+    );
+
+    return uidsForDelete;
+  }
+
+  private async callPartialDelete(uids: string[]) {
+    this.log.debug('CleanupFirestoreHelper.callPartialDelete.start', {
+      uidsForDelete: uids.length,
+    });
+    if (this.dryRun) {
+      this.log.debug('CleanupFirestoreHelper.callPartialDelete.skipped', {
+        dryRun: this.dryRun,
+        uidNumbers: uids.length,
+      });
+
+      return;
+    }
+
+    await this.cloudTaskQueue.addAll(
+      uids.map(
+        (uid) => () =>
+          this.accountDeleteManager.enqueue({
+            uid,
+            reason: AccountDeleteReasonsMap.cleanup,
+          })
+      )
+    );
```

Comment:
No rate limiting needed here, and I think the mysql db will be fine with our query load. Google recommends starting with a max of 500 operations per second to a 'cold' queue and we can then increase it. I'd be impressed if we fetch Firestore documents, then do our mysql query.... fast enough to end up pushing even 500/sec to google though. There'll be pauses as we fetch the batch from Firestore, then hit MySQL, before queuing another batch.

Reply:
+500/sec...challenge accepted! (j/k)

I'll just replace it with a `Promise.all(...)`.  :)